Adopt state_dict_pre_hook in FSDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90625
* __->__ #90436

Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks. This removes `def state_dict` entirely and paves the path for composable API as well.

rohan-varma committed:
Adopt state_dict_pre_hook in FSDP

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Adopt state_dict_pre_hook in FSDP"




[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

rohan-varma committed:
Update on "Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

rohan-varma committed:
Update on "Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

rohan-varma committed:
Update on "Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

rohan-varma committed:
Update on "Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks.



[ghstack-poisoned]

fegin reviewed (approved):
There are some linter errors. Otherwise it is good to land.

rohan-varma committed:
Update on "Adopt state_dict_pre_hook in FSDP"


Use register_state_dict_pre_hook in FSDP to simplify state_dict implementations & remove hacks. This removes `def state_dict` entirely and paves the path for composable API as well.



[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI passed"

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -83,6 +83,7 @@
 )
 from ._state_dict_utils import (
     _post_load_state_dict_hook,
+    _pre_state_dict_hook,
```
It looks like this is going to get sorted in the next `ufmt`.
